pypsa/optimization/constraints.py

```diff
@@ -1688,25 +1694,79 @@ def define_loss_constraints(
     loss = n.model[f"{c.name}-loss"]
     flow = n.model[f"{c.name}-s"]
 
-    # Add upper limit constraint
-    n.model.add_constraints(
-        loss <= upper_limit, name=f"{c.name}-loss_upper", mask=active
-    )
+    if mode == "secants":
+        atol = 1
+        rtol = 0.1
+
+        lossy = r_pu_eff > 0  # only for lines with losses
+        target = (s_nom_max * s_max_pu).where(lossy, 0)
+
+        factor_rtol = 2 * (rtol + sqrt(rtol + rtol**2))  # + 1
+        step_atol = where(lossy, 2 * sqrt(atol / r_pu_eff), 0)
+
+        # start at 0 everywhere
+        p_k = zeros_like(r_pu_eff)
+        p_k.name = "p_k"
+        breakpoints = [p_k.copy()]
+        max_iter = 50
+        it = 0
+        while ((p_k < target) & lossy).any():
+            step_rtol = where(lossy, p_k * factor_rtol, 0)
+            step_k = maximum(step_atol, step_rtol)
+            p_k = p_k + step_k
+            breakpoints.append(p_k.copy())
+            it += 1
+            if it >= max_iter:
+                raise RuntimeError(
+                    "secant loop hit max_iter; check atol/rtol or inputs; current inputs would result in more than 100 additional constraints"
+                )
+
+        # The final step can be made smaller by setting it to the target
+        # This improves the approximation, however may introduce a dependency on s_nom_max:
+        # breakpoints[-1] = target.max("snapshot").where(lossy, 0)
+        # Therefore we do it only for lines that are non-extendable:
+        breakpoints[-1] = target.max("snapshot").where(
+            lossy & ~is_extendable, breakpoints[-1]
+        )
+
+        breakpoints = concat(breakpoints, dim="secants")  # k=0..K
 
-    # Add linearization constraints for each tangent segment
-    for k in range(1, tangents + 1):
-        # Calculate linearization parameters for segment k
-        p_k = k / tangents * s_max_pu * s_nom_max
-        loss_k = r_pu_eff * p_k**2
-        slope_k = 2 * r_pu_eff * p_k
-        offset_k = loss_k - slope_k * p_k
+        a = breakpoints.isel(secants=slice(None, -1))  # k segments: 0..K-1
+        b = breakpoints.isel(secants=slice(1, None))  # k segments: 1..K
+
+        slope = r_pu_eff * (a + b)
+        offset = -r_pu_eff * a * b
 
         # Add constraints for both positive and negative flow
-        for sign in [-1, 1]:
-            lhs = n.model.linexpr((1, loss), (sign * slope_k, flow))
+        for sign, s in [(-1, "neg"), (1, "pos")]:
+            lhs = n.model.linexpr((1, loss), (sign * slope, flow))
             n.model.add_constraints(
-                lhs >= offset_k, name=f"{c.name}-loss_tangents-{k}-{sign}", mask=active
+                lhs >= offset,
+                name=f"{c.name}-loss_secants-{s}",
+                mask=active,
             )
+    else:
+        # Add upper limit constraint
+        n.model.add_constraints(
+            loss <= upper_limit, name=f"{c.name}-loss_upper", mask=active
+        )
```

Comment:
I think this one should apply to the "secant" mode as well. It just bounds the losses (in case of negative prices). Can just remove out of if-else block.